### PR TITLE
Cls

### PR DESCRIPTION
**Pull request by sofiachiarenza**
Added kernel functions and checked that the N5K $C_\ell$ are recovered. Will add documentation soon.

### CODE REVIEW

**Comment by sofiachiarenza** [comment]
@marcobonici I fixed the unit tests and added the documentation for everything, still trying to figure out the "bad" clustering kernel but a part for that it seems ready to me :)

**Comment by sofiachiarenza** [comment]
@marcobonici Hello, now it's ready for real!!!

**Comment by marcobonici** [comment]
Ok!

**Comment by marcobonici** [comment]
(I am gonna be so annoying while reviewing this one, @sofiachiarenza )

File: src/background.jl [hunk 1]
```diff
@@ -90,88 +90,90 @@ over the redshift range specified by the `CosmologicalGrid`.
...
+    for z_idx in 1:length(grid.z_range)
         # Compute the Hubble parameter H(z)
-        BackgroundQuantities.Hz_array[z_idx] = compute_hubble_factor(
-            CosmologicalGrid.z_range[z_idx], AbstractCosmology)
+        bkgq.Hz_array[z_idx] = compute_hubble_factor(grid.z_range[z_idx], cosmo)
         
         # Compute the comoving distance χ(z)
-        BackgroundQuantities.χz_array[z_idx] = compute_χ(
-            CosmologicalGrid.z_range[z_idx], AbstractCosmology)
+        bkgq.χz_array[z_idx] = compute_χ(grid.z_range[z_idx], cosmo)
```
Comment (marcobonici) [inline src/background.jl:100]: This part is fine now, but we are gonna have to modify later. First, you are not using vectorization. Second, you are mutating arrays, which does not work nicely with `Zygote.jl`.  Add a TODO.

File: src/background.jl [hunk 2]
```diff
@@ -187,29 +189,23 @@ Computes the CMB lensing kernel and stores it in the `CMBLensingKernel` struct.
...
-
-
-
-
-
+    n_bins = size(Probe.Kernel, 1)
 
+    if n_bins > 1
+        throw(DomainError("CMB Lensing must have a single tomographic bin!"))
+    end
```
Comment (marcobonici) [inline src/background.jl:205]: Maybe we can enforce this, with a struct redefinition (see `cosmo.jl`).

File: src/cosmo.jl [hunk 3]
```diff
@@ -87,35 +87,65 @@ end
...
+# Parameters
+- `Kernel::AbstractArray{T, 2}`: A 2D array of type `T`, with dimensions `(n_bins, nχ)`. Note that CMB Lensing by definition only has a single tomographic bin.
+
+# Constructors
+- `CMBLensingKernel{T}(n_bins::Int, nχ::Int)`: Creates a `CMBLensingKernel` with the specified `n_bins` and `nχ` values, initializing the kernel values to zeros of type `T`.
+- `CMBLensingKernel(n_bins::Int, nχ::Int)`: Creates a `CMBLensingKernel` with the specified `n_bins` and `nχ` values, initializing the kernel values to zeros of type `Float64`.
 """
 @kwdef mutable struct CMBLensingKernel{T} <: AbstractCosmologicalProbes{T}
-    Kernel::Vector{T} = zeros(500)
+    Kernel::AbstractArray{T, 2} = zeros(1, 1)
```
Comment (marcobonici) [inline src/cosmo.jl:142]: Is there a reason for this being a 2D array? Can't we force this to be a 1D array and then write a slightly different rule when computing, for instance, the $C_\ell$'s?

File: src/integrals.jl [hunk 4]
```diff
@@ -1,3 +1,221 @@
...
+    get_kernel_array(AbstractCosmologicalProbes::GalaxyKernel, 
+        BackgroundQuantities::BackgroundQuantities, R::Vector{T}) where T
+
+Obtains the kernel array for the `GalaxyKernel` probe, with dimensions (bins, nχ, nR).
+
+# Arguments
+- `AbstractCosmologicalProbes::GalaxyKernel`: An instance of the `GalaxyKernel` type.
+- `BackgroundQuantities::BackgroundQuantities`: An instance of the `BackgroundQuantities` type.
+- `R::Vector{T}`: A vector of values for which the kernel array is to be computed.
+"""
```
Comment (marcobonici) [inline src/integrals.jl:50]: Include some LaTeX in the docstring, with the equation.

Comment (marcobonici) [reply src/integrals.jl:50]: True for all the other functions as well, when applicable.

File: test/runtests.jl [hunk 5]
```diff
@@ -254,14 +253,95 @@ end
...
+    cosmo = Blast.FlatΛCDM()
+    bg = Blast.BackgroundQuantities(Hz_array = zeros(200), χz_array=χ )
+
+    cl_test = Blast.compute_Cℓ(pmd, Probe1, Probe2, bg, R, Float64[1.0]) 
+    cl_true = 4950*(R[end]-R[1]) * 2 / π * 2 #2/pi is the ell prefactor, the other 2 comes from the window combination!
 
     @test isapprox(cl_test[1,1,1], cl_true, rtol = 1e-5) 
+end
+
+@testset "More outer integral tests" begin
```
Comment (marcobonici) [inline test/runtests.jl:275]: Why two separate sections?

File: src/integrals.jl [hunk 6]
```diff
@@ -41,44 +259,98 @@ function simpson_weight_array(n::Int; T=Float64)
...
+function compute_Cℓ(w::AbstractArray{T, 3}, K::AbstractArray{T, 4}, bkgq::BackgroundQuantities, weights_χ::AbstractArray{T, 1},
+                    weights_R::AbstractArray{T, 1}, ell_prefactor::AbstractArray{T,1}) where T
+    
+    nχ = length(bkgq.χz_array)
+    Δχ = ((last(bkgq.χz_array)-first(bkgq.χz_array))/(nχ-1))
 
-end
\ No newline at end of file
+    @tullio Cℓ[l,i,j] := ell_prefactor[l]*bkgq.χz_array[n]*K[i,j,n,m]*w[l,n,m]*weights_χ[n]*weights_R[m]*Δχ
```
Comment (marcobonici) [inline src/integrals.jl:353]: I am not sure this is the best way to proceed.
In this case, you are passing the kernel (i.e. the product of the two window function) in order to compute the Cls. You might as well have passed the two windows `WA` and `WB` and written
`@tullio Cℓ[l,i,j] := ell_prefactor[l]*bkgq.χz_array[n]*WA[i,n]*WB[j,m]*w[l,n,m]*weights_χ[n]*weights_R[m]*Δχ`
The advantage is that we don't have to create and store an additional matrix, `K`. We can discuss it, if my suggestion is not ideal.

**Review by marcobonici — COMMENTED** [review]
In general, I am happy with the PR. I am asking you to:
1. Add some LaTeX in the docstrings, to document the physics we are implementing.
2. Add some TODO related to AutoDiff
3. Check whether the computation of the kernel is necessary.
4. Minor stuff

File: src/cosmo.jl (hunk 3 above)
Comment (sofiachiarenza) [reply src/cosmo.jl]: I didn't change the $C_\ell$ computation but how the kernels are handled before.

**Comment by sofiachiarenza** [comment]
@marcobonici blast_paper repo available to check $C_\ell$'s :)

File: src/background.jl [hunk 7]
```diff
@@ -129,13 +129,14 @@ W_g(\\chi) = \\frac{H(z)}{c}n(z)
 
 # Parameters:
 - `nz`: A 2D array of type `T` where each row represents the redshift distribution of galaxies for a specific redshift bin.
+- `z`: The redshift grid corresponding to the nz array.
```
Comment (marcobonici) [inline src/background.jl:132]: I am begin annoying. nz should be written as `nz` as it refers to an array,

Comment (marcobonici) [inline src/background.jl:132]:
```suggestion
- `z`: The redshift grid corresponding to the `nz` array.
```

**Review by marcobonici — COMMENTED** [review]
I am asking for one more modifications (just accept my suggestion).

**Pull request merged**